reader.go

```diff
@@ -30,48 +30,58 @@ func (o *ReaderOptions) norm(name string) {
 
 // Reader reads data from a remote feed.
 type Reader struct {
-	remote *bfs.Object
-	opt    ReaderOptions
-	ctx    context.Context
-	num    int
+	io.Reader
 
-	br io.ReadCloser // bfs reader
-	cr io.ReadCloser // compression reader
-	fd FormatDecoder
+	remotes []*bfs.Object
+	closers []io.Closer
+	fd      FormatDecoder
+	opt     ReaderOptions
+
+	ctx context.Context
+	num int
 }
 
 // NewReader inits a new reader.
 func NewReader(ctx context.Context, remote *bfs.Object, opt *ReaderOptions) (*Reader, error) {
+	return MultiReader(ctx, []*bfs.Object{remote}, opt), nil
+}
+
+// MultiReader inits a new reader for multiple remotes.  Remotes are read sequentially as if concatenated.
+// Once all remotes are fully read, Read will return EOF.
+// MultiReader utilises a single format converter therefore all remotes must be same format.
+func MultiReader(ctx context.Context, remotes []*bfs.Object, opt *ReaderOptions) *Reader {
 	var o ReaderOptions
 	if opt != nil {
 		o = *opt
 	}
-	o.norm(remote.Name())
+	if len(remotes) > 0 {
+		o.norm(remotes[0].Name())
+	}
+	readers := make([]io.Reader, 0, len(remotes))
+	closers := make([]io.Closer, 0, len(remotes))
+	for _, remote := range remotes {
+		r := &streamReader{
+			remote: remote,
+			opt:    o,
+			ctx:    ctx,
+		}
+		readers = append(readers, r)
+		closers = append(closers, r)
+	}
 
 	return &Reader{
-		remote: remote,
-		opt:    o,
-		ctx:    ctx,
-	}, nil
-}
-
-// Read reads raw bytes from the feed.
-func (r *Reader) Read(p []byte) (int, error) {
-	if err := r.ensureOpen(); err != nil {
-		return 0, err
+		Reader:  io.MultiReader(readers...),
+		closers: closers,
+		opt:     o,
+		remotes: remotes,
+		ctx:     ctx,
 	}
-
-	return r.cr.Read(p)
 }
 
 // Decode decodes the next formatted value from the feed.
 func (r *Reader) Decode(v interface{}) error {
-	if err := r.ensureOpen(); err != nil {
-		return err
-	}
-
 	if r.fd == nil {
```

Comment:
this should move to `streamReader`, not sure how this will work with `MultiReader`, I assume you will need some custom `multiDecoder`?

Reply:
I think supporting different-format objects is an overkill. Don't think there will be much need for that at all - sounds synthetic to me.

I'd suggest to just validate objects to be of the same format/compression in constructor to enforce single-format.

Reply:
this could work ...

Reply:
compression is not a problem, each file needs to be opened and decompressed individually anyway, but format is